{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "inpath = './data/ML4H2018/txt/'\n",
    "txt_names = [f for f in os.listdir(inpath) if f.endswith('.txt')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_lower_case(lines):\n",
    "    lines = [l.lower() for l in lines]\n",
    "    return lines\n",
    "\n",
    "def strip_special_chars(lines):\n",
    "    lines = [re.sub('[^a-z]',' ',l) for l in lines]\n",
    "    lines = [re.sub(r'\\s+',' ',l) for l in lines]\n",
    "    return lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_files = len(txt_names)\n",
    "papers_text = {}\n",
    "for i,fname in enumerate(txt_names):\n",
    "    path = os.path.join(inpath, fname)\n",
    "    with open(path, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        \n",
    "        # Strip the references\n",
    "        reference_line = [l for l, line in enumerate(lines) if line == 'References\\n']\n",
    "        if len(reference_line) > 0:\n",
    "            lines = lines[:reference_line[-1]]\n",
    "            \n",
    "        # Add other text pre-processing here.\n",
    "        # \n",
    "        #\n",
    "        lines = to_lower_case(lines)\n",
    "        lines = strip_special_chars(lines)\n",
    "        \n",
    "    paper = \"\".join(lines)\n",
    "    papers_text[fname] = paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nipsml4h2018_paper_103.txt'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(papers_text.keys())[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['interlacing',\n",
       " 'personal',\n",
       " 'and',\n",
       " 'reference',\n",
       " 'genomes',\n",
       " 'for',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'disease',\n",
       " 'variant',\n",
       " 'detection',\n",
       " 'anonymous',\n",
       " 'author',\n",
       " 's',\n",
       " 'affiliation',\n",
       " 'address',\n",
       " 'email',\n",
       " 'abstract',\n",
       " 'dna',\n",
       " 'sequencing',\n",
       " 'to',\n",
       " 'identify',\n",
       " 'genetic',\n",
       " 'variants',\n",
       " 'is',\n",
       " 'becoming',\n",
       " 'increasingly',\n",
       " 'valuable',\n",
       " 'in',\n",
       " 'clinical',\n",
       " 'settings',\n",
       " 'assessment',\n",
       " 'of',\n",
       " 'variants',\n",
       " 'and',\n",
       " 'mutations',\n",
       " 'in',\n",
       " 'such',\n",
       " 'sequencing',\n",
       " 'data',\n",
       " 'is',\n",
       " 'commonly',\n",
       " 'implemented',\n",
       " 'through',\n",
       " 'bayesian',\n",
       " 'heuristic',\n",
       " 'algorithms',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'has',\n",
       " 'shown',\n",
       " 'great',\n",
       " 'promise',\n",
       " 'in',\n",
       " 'improving',\n",
       " 'on',\n",
       " 'these',\n",
       " 'variant',\n",
       " 'calls',\n",
       " 'but',\n",
       " 'the',\n",
       " 'input',\n",
       " 'for',\n",
       " 'these',\n",
       " 'is',\n",
       " 'still',\n",
       " 'a',\n",
       " 'standardized',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'image',\n",
       " 'which',\n",
       " 'is',\n",
       " 'not',\n",
       " 'always',\n",
       " 'best',\n",
       " 'suited',\n",
       " 'to',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'algorithms',\n",
       " 'in',\n",
       " 'this',\n",
       " 'paper',\n",
       " 'we',\n",
       " 'present',\n",
       " 'a',\n",
       " 'novel',\n",
       " 'method',\n",
       " 'for',\n",
       " 'generating',\n",
       " 'images',\n",
       " 'from',\n",
       " 'dna',\n",
       " 'sequencing',\n",
       " 'data',\n",
       " 'which',\n",
       " 'interlaces',\n",
       " 'the',\n",
       " 'human',\n",
       " 'reference',\n",
       " 'genome',\n",
       " 'with',\n",
       " 'personalized',\n",
       " 'sequencing',\n",
       " 'output',\n",
       " 'to',\n",
       " 'maximize',\n",
       " 'usage',\n",
       " 'of',\n",
       " 'sequencing',\n",
       " 'reads',\n",
       " 'and',\n",
       " 'improve',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'algorithm',\n",
       " 'performance',\n",
       " 'we',\n",
       " 'demonstrate',\n",
       " 'the',\n",
       " 'success',\n",
       " 'of',\n",
       " 'this',\n",
       " 'in',\n",
       " 'improving',\n",
       " 'germline',\n",
       " 'variant',\n",
       " 'calling',\n",
       " 'we',\n",
       " 'also',\n",
       " 'furthered',\n",
       " 'this',\n",
       " 'approach',\n",
       " 'to',\n",
       " 'include',\n",
       " 'somatic',\n",
       " 'variant',\n",
       " 'calling',\n",
       " 'across',\n",
       " 'tumour',\n",
       " 'normal',\n",
       " 'paired',\n",
       " 'data',\n",
       " 'by',\n",
       " 'utilizing',\n",
       " 'siamese',\n",
       " 'networks',\n",
       " 'we',\n",
       " 'hope',\n",
       " 'these',\n",
       " 'will',\n",
       " 'be',\n",
       " 'utilized',\n",
       " 'in',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'applications',\n",
       " 'on',\n",
       " 'sequencing',\n",
       " 'data',\n",
       " 'for',\n",
       " 'improved',\n",
       " 'clinical',\n",
       " 'outcomes',\n",
       " 'introduction',\n",
       " 'genetic',\n",
       " 'aberrations',\n",
       " 'are',\n",
       " 'defects',\n",
       " 'in',\n",
       " 'dna',\n",
       " 'which',\n",
       " 'have',\n",
       " 'been',\n",
       " 'linked',\n",
       " 'in',\n",
       " 'many',\n",
       " 'cases',\n",
       " 'to',\n",
       " 'human',\n",
       " 'ill',\n",
       " 'health',\n",
       " 'heritable',\n",
       " 'germline',\n",
       " 'variants',\n",
       " 'have',\n",
       " 'been',\n",
       " 'shown',\n",
       " 'to',\n",
       " 'be',\n",
       " 'directly',\n",
       " 'responsible',\n",
       " 'for',\n",
       " 'many',\n",
       " 'genetic',\n",
       " 'disorders',\n",
       " 'and',\n",
       " 'disease',\n",
       " 'predispositions',\n",
       " 'worldwide',\n",
       " 'in',\n",
       " 'both',\n",
       " 'simpler',\n",
       " 'monogenic',\n",
       " 'inheritance',\n",
       " 'disorders',\n",
       " 'such',\n",
       " 'as',\n",
       " 'cystic',\n",
       " 'fibrosis',\n",
       " 'sickle',\n",
       " 'cell',\n",
       " 'anaemia',\n",
       " 'and',\n",
       " 'thalassaemia',\n",
       " 'and',\n",
       " 'in',\n",
       " 'complex',\n",
       " 'polygenic',\n",
       " 'inheritance',\n",
       " 'disorders',\n",
       " 'such',\n",
       " 'as',\n",
       " 'alzheimer',\n",
       " 's',\n",
       " 'and',\n",
       " 'autoimmune',\n",
       " 'diseases',\n",
       " 'additionally',\n",
       " 'mutations',\n",
       " 'in',\n",
       " 'a',\n",
       " 'humans',\n",
       " 'post',\n",
       " 'conception',\n",
       " 'i',\n",
       " 'e',\n",
       " 'a',\n",
       " 'somatic',\n",
       " 'variant',\n",
       " 'can',\n",
       " 'also',\n",
       " 'cause',\n",
       " 'disease',\n",
       " 'cancer',\n",
       " 'is',\n",
       " 'caused',\n",
       " 'by',\n",
       " 'such',\n",
       " 'somatic',\n",
       " 'mutations',\n",
       " 'and',\n",
       " 'is',\n",
       " 'common',\n",
       " 'worldwide',\n",
       " 'accounting',\n",
       " 'for',\n",
       " 'an',\n",
       " 'estimated',\n",
       " 'million',\n",
       " 'deaths',\n",
       " 'in',\n",
       " 'in',\n",
       " 'clinical',\n",
       " 'settings',\n",
       " 'massively',\n",
       " 'parallel',\n",
       " 'personalized',\n",
       " 'dna',\n",
       " 'sequencing',\n",
       " 'is',\n",
       " 'becoming',\n",
       " 'commonplace',\n",
       " 'with',\n",
       " 'genomics',\n",
       " 'england',\n",
       " 'for',\n",
       " 'example',\n",
       " 'pledging',\n",
       " 'to',\n",
       " 'sequence',\n",
       " 'five',\n",
       " 'million',\n",
       " 'britons',\n",
       " 'over',\n",
       " 'the',\n",
       " 'next',\n",
       " 'five',\n",
       " 'years',\n",
       " 'such',\n",
       " 'efforts',\n",
       " 'will',\n",
       " 'enable',\n",
       " 'broad',\n",
       " 'coverage',\n",
       " 'of',\n",
       " 'the',\n",
       " 'genome',\n",
       " 'for',\n",
       " 'identification',\n",
       " 'of',\n",
       " 'rare',\n",
       " 'variants',\n",
       " 'similarly',\n",
       " 'the',\n",
       " 'sequencing',\n",
       " 'of',\n",
       " 'cancers',\n",
       " 'is',\n",
       " 'gaining',\n",
       " 'popularity',\n",
       " 'due',\n",
       " 'to',\n",
       " 'their',\n",
       " 'genetic',\n",
       " 'complexity',\n",
       " 'and',\n",
       " 'heterogeneity',\n",
       " 'with',\n",
       " 'any',\n",
       " 'given',\n",
       " 'tumor',\n",
       " 'being',\n",
       " 'genetically',\n",
       " 'unique',\n",
       " 'this',\n",
       " 'personalized',\n",
       " 'sequencing',\n",
       " 'information',\n",
       " 'can',\n",
       " 'improve',\n",
       " 'prevention',\n",
       " 'diagnosis',\n",
       " 'and',\n",
       " 'treatment',\n",
       " 'genome',\n",
       " 'sequencing',\n",
       " 'is',\n",
       " 'a',\n",
       " 'rich',\n",
       " 'source',\n",
       " 'of',\n",
       " 'data',\n",
       " 'and',\n",
       " 'a',\n",
       " 'clear',\n",
       " 'target',\n",
       " 'for',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'approaches',\n",
       " 'and',\n",
       " 'algorithms',\n",
       " 'traditionally',\n",
       " 'assessment',\n",
       " 'of',\n",
       " 'variants',\n",
       " 'and',\n",
       " 'mutations',\n",
       " 'in',\n",
       " 'dna',\n",
       " 'sequencing',\n",
       " 'data',\n",
       " 'is',\n",
       " 'commonly',\n",
       " 'implemented',\n",
       " 'through',\n",
       " 'bayesian',\n",
       " 'heuristic',\n",
       " 'algorithms',\n",
       " 'more',\n",
       " 'recently',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'approaches',\n",
       " 'have',\n",
       " 'started',\n",
       " 'to',\n",
       " 'demonstrate',\n",
       " 'improved',\n",
       " 'results',\n",
       " 'for',\n",
       " 'example',\n",
       " 'deepvariant',\n",
       " 'an',\n",
       " 'analysis',\n",
       " 'pipeline',\n",
       " 'developed',\n",
       " 'by',\n",
       " 'google',\n",
       " 'recently',\n",
       " 'won',\n",
       " 'the',\n",
       " 'precisionfda',\n",
       " 'truth',\n",
       " 'challenge',\n",
       " 'through',\n",
       " 'neural',\n",
       " 'network',\n",
       " 'detection',\n",
       " 'of',\n",
       " 'germline',\n",
       " 'variants',\n",
       " 'these',\n",
       " 'approaches',\n",
       " 'rely',\n",
       " 'on',\n",
       " 'a',\n",
       " 'generated',\n",
       " 'image',\n",
       " 'of',\n",
       " 'the',\n",
       " 'sequencing',\n",
       " 'reads',\n",
       " 'known',\n",
       " 'as',\n",
       " 'a',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'image',\n",
       " 'pile',\n",
       " 'ups',\n",
       " 'are',\n",
       " 'commonly',\n",
       " 'produced',\n",
       " 'in',\n",
       " 'bioinformatics',\n",
       " 'research',\n",
       " 'to',\n",
       " 'visualize',\n",
       " 'read',\n",
       " 'alignments',\n",
       " 'at',\n",
       " 'specific',\n",
       " 'genomic',\n",
       " 'loci',\n",
       " 'for',\n",
       " 'example',\n",
       " 'in',\n",
       " 'the',\n",
       " 'integrative',\n",
       " 'genomics',\n",
       " 'viewer',\n",
       " 'however',\n",
       " 'this',\n",
       " 'may',\n",
       " 'not',\n",
       " 'be',\n",
       " 'best',\n",
       " 'suited',\n",
       " 'to',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'applications',\n",
       " 'populationsubmitted',\n",
       " 'to',\n",
       " 'nd',\n",
       " 'conference',\n",
       " 'on',\n",
       " 'neural',\n",
       " 'information',\n",
       " 'processing',\n",
       " 'systems',\n",
       " 'nips',\n",
       " 'do',\n",
       " 'not',\n",
       " 'distribute',\n",
       " 'scale',\n",
       " 'data',\n",
       " 'like',\n",
       " 'the',\n",
       " 'human',\n",
       " 'reference',\n",
       " 'genome',\n",
       " 'is',\n",
       " 'important',\n",
       " 'in',\n",
       " 'understanding',\n",
       " 'how',\n",
       " 'the',\n",
       " 'individual',\n",
       " 's',\n",
       " 'personalized',\n",
       " 'sequence',\n",
       " 'output',\n",
       " 'differs',\n",
       " 'from',\n",
       " 'the',\n",
       " 'general',\n",
       " 'population',\n",
       " 'this',\n",
       " 'reference',\n",
       " 'genome',\n",
       " 'is',\n",
       " 'not',\n",
       " 'well',\n",
       " 'emphasized',\n",
       " 'in',\n",
       " 'the',\n",
       " 'traditional',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'and',\n",
       " 'is',\n",
       " 'confined',\n",
       " 'to',\n",
       " 'a',\n",
       " 'small',\n",
       " 'portion',\n",
       " 'of',\n",
       " 'the',\n",
       " 'image',\n",
       " 'in',\n",
       " 'this',\n",
       " 'paper',\n",
       " 'we',\n",
       " 'describe',\n",
       " 'our',\n",
       " 'efforts',\n",
       " 'to',\n",
       " 'adapt',\n",
       " 'dna',\n",
       " 'sequencing',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'generation',\n",
       " 'to',\n",
       " 'build',\n",
       " 'an',\n",
       " 'improved',\n",
       " 'input',\n",
       " 'for',\n",
       " 'machine',\n",
       " 'learning',\n",
       " 'algorithms',\n",
       " 'we',\n",
       " 'do',\n",
       " 'this',\n",
       " 'by',\n",
       " 'interlacing',\n",
       " 'reference',\n",
       " 'genome',\n",
       " 'sequence',\n",
       " 'with',\n",
       " 'personalized',\n",
       " 'sequencing',\n",
       " 'output',\n",
       " 'to',\n",
       " 'produce',\n",
       " 'a',\n",
       " 'spliced',\n",
       " 'image',\n",
       " 'firstly',\n",
       " 'we',\n",
       " 'demonstrate',\n",
       " 'the',\n",
       " 'value',\n",
       " 'of',\n",
       " 'this',\n",
       " 'through',\n",
       " 'our',\n",
       " 'own',\n",
       " 'germline',\n",
       " 'variant',\n",
       " 'pipeline',\n",
       " 'germlinenet',\n",
       " 'for',\n",
       " 'which',\n",
       " 'the',\n",
       " 'spliced',\n",
       " 'kernal',\n",
       " 'exhibits',\n",
       " 'improved',\n",
       " 'results',\n",
       " 'secondly',\n",
       " 'we',\n",
       " 'incorporate',\n",
       " 'this',\n",
       " 'into',\n",
       " 'a',\n",
       " 'new',\n",
       " 'architecture',\n",
       " 'for',\n",
       " 'somatic',\n",
       " 'tumor',\n",
       " 'variant',\n",
       " 'calling',\n",
       " 'somaticnet',\n",
       " 'which',\n",
       " 'utilizes',\n",
       " 'a',\n",
       " 'novel',\n",
       " 'method',\n",
       " 'of',\n",
       " 'analyzing',\n",
       " 'sequencing',\n",
       " 'information',\n",
       " 'from',\n",
       " 'both',\n",
       " 'tumor',\n",
       " 'tissues',\n",
       " 'and',\n",
       " 'normal',\n",
       " 'tissues',\n",
       " 'of',\n",
       " 'the',\n",
       " 'individual',\n",
       " 'through',\n",
       " 'siamese',\n",
       " 'networks',\n",
       " 'for',\n",
       " 'germline',\n",
       " 'variant',\n",
       " 'data',\n",
       " 'we',\n",
       " 'used',\n",
       " 'one',\n",
       " 'of',\n",
       " 'the',\n",
       " 'platinum',\n",
       " 'genomes',\n",
       " 'a',\n",
       " 'series',\n",
       " 'of',\n",
       " 'well',\n",
       " 'studied',\n",
       " 'genomes',\n",
       " 'originally',\n",
       " 'sequenced',\n",
       " 'as',\n",
       " 'part',\n",
       " 'of',\n",
       " 'the',\n",
       " 'genomes',\n",
       " 'project',\n",
       " 'this',\n",
       " 'included',\n",
       " 'serial',\n",
       " 'sequencing',\n",
       " 'of',\n",
       " 'immediate',\n",
       " 'family',\n",
       " 'members',\n",
       " 'to',\n",
       " 'enable',\n",
       " 'haplotype',\n",
       " 'reconstruction',\n",
       " 'and',\n",
       " 'high',\n",
       " 'confidence',\n",
       " 'variant',\n",
       " 'identification',\n",
       " 'the',\n",
       " 'member',\n",
       " 'ceph',\n",
       " 'pedigree',\n",
       " 'was',\n",
       " 'sequenced',\n",
       " 'to',\n",
       " 'x',\n",
       " 'depth',\n",
       " 'on',\n",
       " 'a',\n",
       " 'hiseq',\n",
       " 'system',\n",
       " 'of',\n",
       " 'which',\n",
       " 'we',\n",
       " 'utilized',\n",
       " 'the',\n",
       " 'well',\n",
       " 'characterized',\n",
       " 'na',\n",
       " 'sample',\n",
       " 'for',\n",
       " 'somatic',\n",
       " 'variant',\n",
       " 'detection',\n",
       " 'we',\n",
       " 'utilized',\n",
       " 'a',\n",
       " 'dataset',\n",
       " 'generated',\n",
       " 'through',\n",
       " 'bamsurgeon',\n",
       " 'a',\n",
       " 'tool',\n",
       " 'designed',\n",
       " 'to',\n",
       " 'spike',\n",
       " 'in',\n",
       " 'artificial',\n",
       " 'mutations',\n",
       " 'in',\n",
       " 'a',\n",
       " 'sequencing',\n",
       " 'file',\n",
       " 'somatic',\n",
       " 'variant',\n",
       " 'callers',\n",
       " 'run',\n",
       " 'on',\n",
       " 'a',\n",
       " 'tumor',\n",
       " 'normal',\n",
       " 'pair',\n",
       " 'of',\n",
       " 'files',\n",
       " 'for',\n",
       " 'our',\n",
       " 'data',\n",
       " 'set',\n",
       " 'we',\n",
       " 'utilized',\n",
       " 'a',\n",
       " 'pair',\n",
       " 'of',\n",
       " 'independently',\n",
       " 'sequenced',\n",
       " 'samples',\n",
       " 'from',\n",
       " 'the',\n",
       " 'same',\n",
       " 'individual',\n",
       " 'with',\n",
       " 'bamsurgeon',\n",
       " 'applied',\n",
       " 'to',\n",
       " 'one',\n",
       " 'sample',\n",
       " 'output',\n",
       " 'to',\n",
       " 'add',\n",
       " 'known',\n",
       " 'variants',\n",
       " 'and',\n",
       " 'form',\n",
       " 'a',\n",
       " 'synthetic',\n",
       " 'tumour',\n",
       " 'germlinenet',\n",
       " 'and',\n",
       " 'somaticnet',\n",
       " 'shared',\n",
       " 'a',\n",
       " 'data',\n",
       " 'pre',\n",
       " 'processing',\n",
       " 'pipeline',\n",
       " 'similar',\n",
       " 'to',\n",
       " 'that',\n",
       " 'of',\n",
       " 'google',\n",
       " 's',\n",
       " 'deepvariant',\n",
       " 'which',\n",
       " 'generates',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'images',\n",
       " 'from',\n",
       " 'the',\n",
       " 'raw',\n",
       " 'dna',\n",
       " 'sequencing',\n",
       " 'data',\n",
       " 'this',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'image',\n",
       " 'is',\n",
       " 'then',\n",
       " 'rgb',\n",
       " 'encoded',\n",
       " 'and',\n",
       " 'used',\n",
       " 'as',\n",
       " 'input',\n",
       " 'to',\n",
       " 'the',\n",
       " 'cnns',\n",
       " 'for',\n",
       " 'classification',\n",
       " 'this',\n",
       " 'process',\n",
       " 'was',\n",
       " 'implemented',\n",
       " 'over',\n",
       " 'two',\n",
       " 'modules',\n",
       " 'the',\n",
       " 'first',\n",
       " 'module',\n",
       " 'identifies',\n",
       " 'variant',\n",
       " 'candidate',\n",
       " 'sites',\n",
       " 'using',\n",
       " 'a',\n",
       " 'frequency',\n",
       " 'heuristic',\n",
       " 'the',\n",
       " 'human',\n",
       " 'genome',\n",
       " 'contains',\n",
       " 'billion',\n",
       " 'base',\n",
       " 'pairs',\n",
       " 'in',\n",
       " 'which',\n",
       " 'germline',\n",
       " 'variants',\n",
       " 'occur',\n",
       " 'at',\n",
       " 'approximately',\n",
       " 'one',\n",
       " 'in',\n",
       " 'every',\n",
       " 'bases',\n",
       " 'in',\n",
       " 'order',\n",
       " 'to',\n",
       " 'reduce',\n",
       " 'the',\n",
       " 'number',\n",
       " 'of',\n",
       " 'sites',\n",
       " 'the',\n",
       " 'cnn',\n",
       " 'needs',\n",
       " 'to',\n",
       " 'classify',\n",
       " 'we',\n",
       " 'use',\n",
       " 'an',\n",
       " 'initial',\n",
       " 'filter',\n",
       " 'with',\n",
       " 'high',\n",
       " 'sensitivity',\n",
       " 'and',\n",
       " 'low',\n",
       " 'specificity',\n",
       " 'we',\n",
       " 'used',\n",
       " 'the',\n",
       " 'same',\n",
       " 'simple',\n",
       " 'heuristic',\n",
       " 'variant',\n",
       " 'allele',\n",
       " 'frequency',\n",
       " 'vaf',\n",
       " 'percentage',\n",
       " 'of',\n",
       " 'mismatched',\n",
       " 'bases',\n",
       " 'given',\n",
       " 'a',\n",
       " 'location',\n",
       " 'threshold',\n",
       " 'of',\n",
       " 'following',\n",
       " 'deepvariant',\n",
       " 's',\n",
       " 'approach',\n",
       " 'the',\n",
       " 'second',\n",
       " 'module',\n",
       " 'creates',\n",
       " 'rgb',\n",
       " 'encoded',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'images',\n",
       " 'for',\n",
       " 'each',\n",
       " 'candidate',\n",
       " 'site',\n",
       " 'base',\n",
       " 'sequences',\n",
       " 'from',\n",
       " 'reads',\n",
       " 'overlapping',\n",
       " 'candidate',\n",
       " 'sites',\n",
       " 'identified',\n",
       " 'earlier',\n",
       " 'are',\n",
       " 'encoded',\n",
       " 'as',\n",
       " 'tensor',\n",
       " 'objects',\n",
       " 'in',\n",
       " 'the',\n",
       " 'first',\n",
       " 'channel',\n",
       " 'a',\n",
       " 'second',\n",
       " 'channel',\n",
       " 'is',\n",
       " 'used',\n",
       " 'to',\n",
       " 'encode',\n",
       " 'base',\n",
       " 'quality',\n",
       " 'scores',\n",
       " 'in',\n",
       " 'theory',\n",
       " 'additional',\n",
       " 'channels',\n",
       " 'can',\n",
       " 'be',\n",
       " 'added',\n",
       " 'to',\n",
       " 'represent',\n",
       " 'information',\n",
       " 'from',\n",
       " 'the',\n",
       " 'sequencing',\n",
       " 'process',\n",
       " 'such',\n",
       " 'as',\n",
       " 'dna',\n",
       " 'strand',\n",
       " 'positive',\n",
       " 'or',\n",
       " 'negative',\n",
       " 'and',\n",
       " 'mapping',\n",
       " 'quality',\n",
       " 'scores',\n",
       " 'mapqs',\n",
       " 'the',\n",
       " 'purpose',\n",
       " 'of',\n",
       " 'the',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'image',\n",
       " 'is',\n",
       " 'to',\n",
       " 'allow',\n",
       " 'a',\n",
       " 'direct',\n",
       " 'comparison',\n",
       " 'of',\n",
       " 'dna',\n",
       " 'fragments',\n",
       " 'from',\n",
       " 'the',\n",
       " 'dataset',\n",
       " 'with',\n",
       " 'the',\n",
       " 'reference',\n",
       " 'to',\n",
       " 'identify',\n",
       " 'variants',\n",
       " 'traditionally',\n",
       " 'a',\n",
       " 'pile',\n",
       " 'up',\n",
       " 'displays',\n",
       " 'a',\n",
       " 'section',\n",
       " 'of',\n",
       " 'the',\n",
       " 'human',\n",
       " 'reference',\n",
       " 'genome',\n",
       " 'along',\n",
       " 'the',\n",
       " 'top',\n",
       " 'five',\n",
       " 'pixels',\n",
       " 'of',\n",
       " 'the',\n",
       " 'image',\n",
       " 'together',\n",
       " 'with',\n",
       " 'the',\n",
       " 'sequenced',\n",
       " 'fragments',\n",
       " 'of',\n",
       " 'dna',\n",
       " 'from',\n",
       " 'the',\n",
       " 'given',\n",
       " 'dataset',\n",
       " 'that',\n",
       " 'align',\n",
       " 'to',\n",
       " 'this',\n",
       " 'region',\n",
       " 'however',\n",
       " 'as',\n",
       " 'the',\n",
       " 'cnn',\n",
       " ...]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papers_text['nipsml4h2018_paper_103.txt'].split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "interlacing personal and reference genomes for machine learning disease variant detection  anonymous author s affiliation address email  abstract dna sequencing to identify genetic variants is becoming increasingly valuable in clinical settings assessment of variants and mutations in such sequencing data is commonly implemented through bayesian heuristic algorithms machine learning has shown great promise in improving on these variant calls but the input for these is still a standardized pile up image which is not always best suited to machine learning algorithms in this paper we present a novel method for generating images from dna sequencing data which interlaces the human reference genome with personalized sequencing output to maximize usage of sequencing reads and improve machine learning algorithm performance we demonstrate the success of this in improving germline variant calling we also furthered this approach to include somatic variant calling across tumour normal paired data by utilizing siamese networks we hope these will be utilized in machine learning applications on sequencing data for improved clinical outcomes                                           introduction  genetic aberrations are defects in dna which have been linked in many cases to human ill health heritable germline variants have been shown to be directly responsible for many genetic disorders and disease predispositions worldwide in both simpler monogenic inheritance disorders such as cystic fibrosis sickle cell anaemia and thalassaemia and in complex polygenic inheritance disorders such as alzheimer s and autoimmune diseases additionally mutations in a humans post conception i e a somatic variant can also cause disease cancer is caused by such somatic mutations and is common worldwide accounting for an estimated million deaths in in clinical settings massively parallel personalized dna sequencing is becoming commonplace with genomics england for example pledging to sequence five million britons over the next five years such efforts will enable broad coverage of the genome for identification of rare variants similarly the sequencing of cancers is gaining popularity due to their genetic complexity and heterogeneity with any given tumor being genetically unique this personalized sequencing information can improve prevention diagnosis and treatment genome sequencing is a rich source of data and a clear target for machine learning approaches and algorithms traditionally assessment of variants and mutations in dna sequencing data is commonly implemented through bayesian heuristic algorithms more recently machine learning approaches have started to demonstrate improved results for example deepvariant an analysis pipeline developed by google recently won the precisionfda truth challenge through neural network detection of germline variants these approaches rely on a generated image of the sequencing reads known as a pile up image pile ups are commonly produced in bioinformatics research to visualize read alignments at specific genomic loci for example in the integrative genomics viewer however this may not be best suited to machine learning applications populationsubmitted to nd conference on neural information processing systems nips do not distribute      scale data like the human reference genome is important in understanding how the individual s personalized sequence output differs from the general population this reference genome is not well emphasized in the traditional pile up and is confined to a small portion of the image    in this paper we describe our efforts to adapt dna sequencing pile up generation to build an improved input for machine learning algorithms we do this by interlacing reference genome sequence with personalized sequencing output to produce a spliced image firstly we demonstrate the value of this through our own germline variant pipeline germlinenet for which the spliced kernal exhibits improved results secondly we incorporate this into a new architecture for somatic tumor variant calling somaticnet which utilizes a novel method of analyzing sequencing information from both tumor tissues and normal tissues of the individual through siamese networks              for germline variant data we used one of the platinum genomes a series of well studied genomes originally sequenced as part of the genomes project this included serial sequencing of immediate family members to enable haplotype reconstruction and high confidence variant identification the member ceph pedigree was sequenced to x depth on a hiseq  system of which we utilized the well characterized na sample for somatic variant detection we utilized a dataset generated through bamsurgeon a tool designed to spike in artificial mutations in a sequencing file somatic variant callers run on a tumor normal pair of files for our data set we utilized a pair of independently sequenced samples from the same individual with bamsurgeon applied to one sample output to add known variants and form a synthetic tumour          germlinenet and somaticnet shared a data pre processing pipeline similar to that of google s deepvariant which generates pile up images from the raw dna sequencing data this pile up image is then rgb encoded and used as input to the cnns for classification this process was implemented over two modules the first module identifies variant candidate sites using a frequency heuristic the human genome contains billion base pairs in which germline variants occur at approximately one in every bases in order to reduce the number of sites the cnn needs to classify we use an initial filter with high sensitivity and low specificity we used the same simple heuristic variant allele frequency vaf percentage of mismatched bases given a location threshold of following deepvariant s approach the second module creates rgb encoded pile up images for each candidate site base sequences from reads overlapping candidate sites identified earlier are encoded as tensor objects in the first channel a second channel is used to encode base quality scores in theory additional channels can be added to represent information from the sequencing process such as dna strand positive or negative and mapping quality scores mapqs        the purpose of the pile up image is to allow a direct comparison of dna fragments from the dataset with the reference to identify variants traditionally a pile up displays a section of the human reference genome along the top five pixels of the image together with the sequenced fragments of dna from the given dataset that align to this region however as the cnn can only extract features which the kernel is applied to the reference dna is then only compared with the normal dna at the border we introduce a novel method for improving pile up image input and maximizing feature use through a pile up design called the spliced pile up the suggested spliced pile up approach combines the reference dna and the sequenced dna for the first layer of the cnn keeping the reference at a constant position in the kernel this allows the kernel to learn the weights needed to appropriately combine the two pieces of information the reference dna and the normal dna we achieve this by altering input images such as to insert the reference at every s rows where s is the vertical stride of the convolutional kernel s an example of these spliced images is shown in figure                                         methods data generation  data pre processing  spliced pile up          germlinenet uses a cnn to classify the difference between the actual base sequence and the reference in pile up images as being a an error due to sequencing or b caused by a germline variant inception and resnet were candidate state of the art network architectures for image classifications  we used the softmax as our final layer to calculate class probabilities and the negative log likelihood as the loss function for gradient descent for model training we used a balanced dataset of pile up images labelled as true positive from ground truth and images from random locations outside of candidate sites training and validation sets were split on the number of chromosomes training testing and x          cnn architecture and training    for somatic variant calling our new approach is based on a siamese cnn somaticnet takes two images as inputs feeds them forward through the neural network to give two vectors in latent space and returns a function applied to the distance between the two vectors we use the euclidean distance between the two vectors and train with the contrastive loss function where distance smaller than a threshold value between the two vectors would indicate a no mutation present in the tumour sample s candidate site while a larger distance would mean b a mutation present in the tumour sample s candidate site the weights from germlinenet are used to initialise those in somaticnet making the two sister networks we used a balanced dataset of images labelled as true positives and a matching number of true negatives outside candidate sites            different model architectures and the addition of the spliced pile up were tested with germlinenet using k fold cross validation k test statistics show the average of the best model of each cross validation cycle using f score table epochs of training was conducted in each cycle and statistics were calculated on the test fold the architecture resnet resulted in the highest f  and using the spliced pile up and quality improved this further to figure                results germlinenet  figure base layer of the normal image and the spliced image the splicing is performed by having the reference at alternating lines the splicing is done for each layer however this image only shows the base to make the example clearer number of epochs vs validation loss of resnet for the first fold of cross validation with the spliced image learns quicker and reaches a higher accuracy       we evaluated somaticnet s performance using similar cross validation procedures with the effects of model architectures and the spliced pile up summarised in table the siamese network using resnet showed better performance with spliced pile up f score it should be noted that including quality scores led to deterioration in model performance potentially because our semi simulated data set had synthetic mutated variants that do not correlate with data quality scores           somaticnet  distribution of siamese distance on validation data showed that somaticnet was able to distinguish between tumour and normal tissue classes figure actual threshold of siamese distance used for somatic variant calling can be optimised through a grid search with high validation accuracy sensitivity and precision figure    model  spliced  resnet  false  model  spliced  resnet  false  quality  accuracy  f  roc auc  recall  precision  false      true      true false      true      inception v false false      true      true false      true      table results of k fold cross validation for germlinenet s neural network s f  quality  accuracy  f  roc auc  recall  precision  false      true      true false      true      inception v false false      true      true false      true      table results of k fold cross validation for somaticnet s neural network s f    tumour normal    validation accuracy  number of images            classicfication accuracy true positive ture negative           siamese distance square root transformed     siamese distance square root transformed    figure siamese distance distribution from tumour and normal tissue class in validation set showing the model was able to learn to distinguish the two validation accuracy vs different siamese distances threshold of distances can be optimised to achieve high sensitivity and specificity when calling variants      conclusion    our approaches provide promising novel machine learning based methods for improving genetic variant identification these are used to both increase accuracy and to enable machine learning analysis of tumour normal pairs in cancers additionally these pile up methods have the potential for use beyond just genomic variant detection for example alternative splicing in transcriptomic pile ups our hope is that these methods will be used to increase the success of machine learning algorithms in sequencing    \n"
     ]
    }
   ],
   "source": [
    "print(papers_text['nipsml4h2018_paper_103.txt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
